### DISCUSSION TITLE
Variable number of parameters for a command

### DISCUSSION THREAD

**Discussion by grbba** [daniele77/cli#93] [Q&A] [answered]
Hello,

first thanks for developing this nice feature and i am interested in using it as a front end to feed commands to an Arduino MCU.  The question I have would be on how to create something like a 'send' command which takes a variable number of arguments depending on the command to send to the MCU. 

In fact the MCU understands ( quite cryptic but for the users that make sense :) ) commands in the form of
< Pi ... Pn > with at least one parameter and at most n (with n at most around 10). The enclosing brackets i can build obviously. Some examples:

On the cli                  ->  send over the wire ( Serial and/or Ethernet ) to the MCU
send R                       -> <R>  
send R 5 3                 -> <R 5 3>
send D ACK ON        -> <D ACK ON>
send T 25 156 444. -> <T 25  156 444>

I don't need the types of the parameters they can all be strings no issue as they will be interpreted on the MCU as needed.

Is there a way of doing this today, which i didn't see yet or maybe an improvement to add variadic parameters in the menu construction ? 
I tried to see if that is possible in your code base to do but i have to admit that i am not that fluent in C++ to fully understand in a reasonable time how your code works to be able to add that .... ( but if you have pointers i may try to do it )

Thanks a lot
//

**Comment by daniele77** [comment, marked as answer]
Hi @grbba .

First of all, thank you for using the CLI library and contributing to the community with your good question. Please add a star to the project and -- if you feel really generous -- you can consider becoming a sponsor by clicking the heart button on top of the page :-)

I moved your question to "GitHub discussions" because it's more appropriate here.

Currently, the library allows adding a command with a variable number of `std::string` parameters. You can see it in the `complete` example that comes with the library:

    rootMenu->Insert(
            "sort",
            [](std::ostream& out, std::vector<std::string> data)
            {
                std::sort(data.begin(), data.end());
                out << "sorted list: ";
                std::copy(data.begin(), data.end(), std::ostream_iterator<std::string>(out, " "));
                out << "\n";
            },
            "Alphabetically sort a list of words" );

So, in your case, you could have something like this:

    rootMenu->Insert(
            "send",
            [](std::ostream& out, std::vector<std::string> data)
            {
                // parse data and do something with it
            },
            "Send a command to MCU" );

However, maybe your design could be improved. It depends on the context of your application.
You can also consider adding a different CLI command for each kind of MCU command you want to send (e.g. `sendR`, `sendD`, `sendT`), if your application must take different actions in the three cases.

**Reply by grbba** [reply]
Thanks a lot :) That works perfectly. I have a remark though and that is a difference between menu items having simple data types and the one just proposed using the vector:

```
 - open <string> <string> <int>
        open <serial|ethernet> <port> <baud>; If serial indicate the used USB port and for ethernet indicate the
        IP address of the commandstation baud will be ignored for ethernet and if not specified for serial
        the deafult of 115200 will be used
 - send
        Send a < > command to the commandstation.
```
as you can see the open command uses three params for which their types are named but not the 'send' command being a vector. Is there some place that cloud be changed to make it consistent ?

**Reply by daniele77** [reply]
The library has multiple versions of the `Insert` function, in overloading.
In particular, you can write one of these:

    // default parameter help
    rootMenu->Insert(
            "mycommand",
            [](std::ostream& out, int par1, double par2, string par3)
            {
                 // do something
            },
            "Help  for the mycommand command" );

    // explicit parameter help
    rootMenu->Insert(
            "mycommand", {"par1 desc", "par2 desc", "par3 desc"},  // <--- here you can describe the parameters
            [](std::ostream& out, int par1, double par2, string par3)
            {
                 // do something
            },
            "Help  for the mycommand command" );

In your case you can write something like this:

    rootMenu->Insert(
        "send",
        {"list of command parameter separated by space"},
        [](std::ostream& out, std::vector<std::string> data)
        {
            // parse data and do something with it
        },
        "Send a command to MCU" );

**Reply by grbba** [reply]
Thanks for your fast reply 👍  That's what i was looking for!
